Distilling whisper message and envelope code into message.go

kortatu opened the pull request:
Removed `message.go` file contained references to whisper message and envelope code. All that whisper code has been copied and distilled into swarm/pss/crypto as the default crypto implementation.
Crypto interface now is composed of 3 interfaces. 
- MessageCrypto (contains message wrapping unwrapping and validation)
- BackendCrypto (contains Key manipulation, storage and conversion)
- CryptoUtils (utilities for tests)

Sealing the message (poW and Nonce) has been completely removed.

We keep the whisper structure of encrypted message with the format:
` flags+payloadSize+payload+padding+signature` 
but only inside the default crypto implementation. Now the payload of the whisper message is a simple `[]byte`.

Topic has been moved from PssMsg.Payload to PssMsg.

kortatu committed:
Distilling whisper message and envelope code into message.go

aperonam commented on pss/crypto.go:
```diff
@@ -17,19 +17,49 @@ package pss
+	"errors"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"io"
+	mrand "math/rand"
```
Is this needed also? or just crypto/rand is enough

aperonam replied:
Should we just use crypto/rand?

aperonam commented on pss/crypto.go:
```diff
@@ -74,21 +186,98 @@ func (crypto *defaultCryptoBackend) AddSymKeyDirect(bytes []byte) (string, error
+	return res, nil
+}
+
+// validateDataIntegrity returns false if the data have the wrong or contains all zeros,
```
```suggestion
// validateDataIntegrity returns false if the data has the wrong size or contains all zeros,
```

kortatu replied:
data is plural

kortatu replied:
Fixed size in commit 86a3cd7

aperonam commented on pss/crypto.go:
```diff
@@ -74,21 +186,98 @@ func (crypto *defaultCryptoBackend) AddSymKeyDirect(bytes []byte) (string, error
+
+// validateDataIntegrity returns false if the data have the wrong or contains all zeros,
+// which is the simplest and the most common bug.
+func (crypto *defaultCryptoBackend) validateDataIntegrity(k []byte, expectedSize int) bool {
```
I think the receiver is not needed

kortatu replied:
Changed validateDataIntegrity to a simple function without receiver. Commit 86a3cd7

kortatu committed:
pss: Moved crypto to own package.
Distilled whisper message structure and encryption to default crypto implementation.
Added Topic to PssMsg
Simplified PssMsg Payload to []byte

kortatu committed:
Fixed pr comments

aperonam commented on pss/notify/notify_test.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"context"
 	"flag"
 	"fmt"
+	crypto2 "github.com/ethersphere/swarm/pss/crypto"
```
Why crypto2?

kortatu replied:
Fixed in 592fe59

aperonam commented on pss/types.go:
```diff
@@ -228,3 +227,16 @@ func BytesToTopic(b []byte) Topic {
 	return toTopic(topicHashFunc.Sum(nil))
 }
+
+// BytesToTopic converts from the byte array representation of a topic
```
```suggestion
// toTopic converts from the byte array representation of a topic
```

kortatu replied:
Fixed in 592fe59

kortatu committed:
pr review changes

kortatu committed:
defaultCryptoBackend should be a pointer in receivedMessage.
Doc added for Crypto

kortatu committed:
pss: PR #1731 comments

kortatu committed:
ss: PR #1731. Comments on public Utils methods.
Removed unused method in crypto.go

kortatu committed:
more pr #1731 comments

kortatu committed:
removed whipser import

kortatu committed:
pss: removed crypto/utils use directly ethereum/crypto to generate private keys in tests

kortatu committed:
pss: returned struct in crypto constructor

kortatu committed:
pss: Renamed SymKey->SymmetricKey. Fixed comments. Removed keccak256 function

kortatu committed:
Merge branch 'master' into issue-1703
